proto/gossip_delegates.go

```diff
@@ -256,11 +316,40 @@ func (gd *GossipDelegate) NotifyAlive(node *memberlist.Node) error {
 	if err == nil && diffNode.Status != types.NODE_STATUS_UP {
 		gd.UpdateNodeStatus(types.NodeId(nodeName), types.NODE_STATUS_UP)
 		gd.triggerStateEvent(types.NODE_ALIVE)
+		if diffNode.Status == types.NODE_STATUS_SUSPECT_DOWN {
+			// Remove the node from probation list
+			logrus.Infof("gossip: Node %v is no more suspected as offline", nodeName)
+			if err := gd.nodeDownProbationManager.Remove(gd.nodeNameToProbationID(nodeName)); err != nil {
+				logrus.Warnf("gossip: Unable to remove suspected down node %v from probation list: %v", nodeName, err)
+			}
+		}
 	} // else if err != nil -> A new node sending us data. We do not add node unless it is added
 	// in our local map externally
 	return nil
 }
 
+func (gd *GossipDelegate) probationExpiredOnSuspectedDownNode(probationID string, nodeData interface{}) error {
+	// Node is suspected to be down for more than the probation timeout
+	// Update the node status to Offline
+	nodeName := gd.probationIDToNodeName(probationID)
+	logrus.Infof("gossip: probation time expired for suspected offline node %v ", nodeName)
+	selfStatus := gd.GetSelfStatus()
+	if selfStatus == types.NODE_STATUS_SUSPECT_NOT_IN_QUORUM {
+		// We are probably out of quorum
+		// Wait again before we mark this node down
+		logrus.Infof("gossip: we are suspected not in quorum, adding suspected offline node %v back to probation list", nodeName)
+		if err := gd.nodeDownProbationManager.Add(probationID, nil, true); err != nil {
+			logrus.Warnf("gossip: Unable to add suspected down node %v from probation list: %v", nodeName, err)
+		}
+		return nil
+	}
```

Comment:
Mark a node down only if a member in the same MetroDomain is alive and you're in quorum

Reply:
done